### DISCUSSION TITLE
How to know if a subscription is still active (websocket alive) or not

### DISCUSSION THREAD

**Discussion by sdekna** [urql-graphql/urql#2780] [Q&A] [answered]
I am using urql with svelte... I subscribe to events but sometimes the websocket is terminated by the server.
Is there a way to know if the subscription is still alive or not? knowing that I would be able to re-subscribe again.

Similar un-answered discussion: https://github.com/FormidableLabs/urql/discussions/2700

**Comment by kitten** [comment, marked as answer]
Ah, yes, this is a little tricky. The tricky part is that `urql` basically does not provide a WebSockets client. Basically, we have no opinion on the transport protocol you use for GraphQL, just like the main spec doesn't. The exception is that obviously we do support GraphQL over HTTP out of the box, but I mean, obviously people expect GraphQL to be an HTTP protocol.

Anyway, if we're taking this as an example: https://github.com/enisdenjo/graphql-ws

It's the WS client's responsibility to make sure that either GraphQL requests are delivered and a result comes back, or that an error is issued. So, if you want to keep track of whether an operation is undeliverable then it's worth checking in your WS client's docs in which cases it bails, issues an error, or retries infinitely

**Reply by JoviDeCroock** [reply]
Also an answer how someone did it can be found in https://github.com/FormidableLabs/urql/discussions/2700

**Comment by sdekna** [comment]
Thanks to @vlastoun for the inspiration, this is how my socket client looks like now:
```js
let socketClient =
	(browser &&
		createWSClient({
			url: "ws://server-url",
			retryAttempts: Infinity,
			shouldRetry: () => true,
			connectionParams: () => {
				// get current access_token, here I am using nhost as my auth provider
				const access_token = nhost.auth.getAccessToken();
				return {
					headers: {
						// 'x-hasura-admin-secret': 'Saif992378579'
						Authorization: `Bearer ${access_token}`,
					},
				};
			},
			on: {
				connected: (socket) => {
					// get the expiry date of the current token and compare it to the time when the socket is connected... then get
					// the difference between and create a timer to close the socket when the token expires.
					// since retryAttempts is set to Infinity, it will try to reconnect again to the socket, and while doing so it
					// will get a new access_token with a new expiration date, and so on...
					const decoded_jwt = nhost.auth.getDecodedAccessToken();
					const expiry_time = decoded_jwt.exp;
					const current_time = Math.round(+new Date() / 1000);
					const difference_time = (expiry_time - current_time) * 1000;
					setTimeout(() => {
						if (socket.readyState === WebSocket.OPEN) {
							socket.close(CloseCode.Forbidden, "Forbidden");
							should_restart.set(true);
						}
					}, difference_time);
				},
				error: (event) => {
					// in case of a fatal error, the re-connection will not work.
					// should_restart is an observable store that is used as as a trigger, when true the frontend framework will
					// kill the current connection and then re-establish a new one... This is to ensure continuouaty inspite
					// of possible fatal errors, especially when using it on a mobile device
					should_restart.set(true);
				},
			},
		})) ||
	null;
```

**Reply by sdekna** [reply]
For reference, I am using capacitor js on top of a sveltekit to generate android/ios apps.
